app/src/main/java/com/example/coctailapp/ui/screens/main/content/shopping/ShoppingListViewModel.kt

```diff
@@ -0,0 +1,131 @@
+    private var _shoppingListByRecipe = MutableStateFlow<List<ShoppingListModel>>(emptyList())
+    val shoppingListByRecipe = _shoppingListByRecipe.asStateFlow()
+    private var _unwantedIngredients = MutableStateFlow<List<ShoppingListIngredient>>(emptyList())
+    val unwantedIngredients = _unwantedIngredients.asStateFlow()
+    private var _showDialog = MutableStateFlow<ShoppingListDialogEvent>(ShoppingListDialogEvent.DialogDismissedEvent)
+    val showDialog = _showDialog.asStateFlow()
+
+
+    fun getUserShoppingList(email: String) {
+
+        shoppingListIngredient = shoppingListDao.getUserShoppingList(email)
+
+        createShoppingListForLazyColumn()
+
+    }
+
+
+    private fun createShoppingListForLazyColumn() {
+        viewModelScope.launch {
+
+            shoppingListIngredient.collectLatest { shoppingListFromFlow ->
+                var previousCocktailId = ""
+                val shoppingList = mutableListOf<ShoppingListModel>()
+                for (shoppingListIngredients in shoppingListFromFlow) {
+                    if (previousCocktailId == shoppingListIngredients.cocktailId) {
+                        shoppingList.last().neededIngredient.add(
+                            Ingredient(
+                                shoppingListIngredients.cocktailIngredientName,
+                                shoppingListIngredients.cocktailIngredientMeasure
+                            )
+                        )
+                    } else {
+                        previousCocktailId = shoppingListIngredients.cocktailId
+                        shoppingList.add(
+                            ShoppingListModel(
+                                shoppingListIngredients.cocktailId,
+                                shoppingListIngredients.cocktailName,
+                                mutableListOf(
+                                    Ingredient(
+                                        shoppingListIngredients.cocktailIngredientName,
+                                        shoppingListIngredients.cocktailIngredientMeasure
+                                    )
+                                )
+                            )
+                        )
+                    }
+                }
+
+                _shoppingListByRecipe.value = shoppingList
+            }
+
+        }
+    }
+
+
+    fun addOrRemoveUnwantedIngredient(ingredientName: String, cocktailId: String, userId: String) {
+        val ingredient = ShoppingListIngredient(
+            userId,
+            cocktailId,
+            "",
+            ingredientName,
+            ""
+        )
+        if (!_unwantedIngredients.value.contains(ingredient))
+            _unwantedIngredients.value += listOf(ingredient)
+        else
+            _unwantedIngredients.value -= listOf(ingredient)
+
+    }
+
+    fun deleteUnwantedIngredients() {
+
+
+        viewModelScope.launch {
+
+
+            unwantedIngredients.value.map { i ->
```

Comment:
Since Room can accept list of objects as a parameter, instead of just one object, in combination with @Delete annotation, we can use that instead of async and map